bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoDatabaseWatchService.java

```diff
@@ -0,0 +1,149 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.ConfigConstants;
+import org.eclipse.smarthome.core.service.AbstractWatchService;
+import org.openhab.binding.miio.internal.MiIoBindingConstants;
+import org.openhab.binding.miio.internal.Utils;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonIOException;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link MiIoDatabaseWatchService} creates a registry of database file per ModelId
+ *
+ * @author Marcel Verpaalen - Initial contribution
+ */
+@Component(service = MiIoDatabaseWatchService.class)
+@NonNullByDefault
+public class MiIoDatabaseWatchService extends AbstractWatchService {
+    private static final String LOCAL_DATABASE_PATH = ConfigConstants.getConfigFolder() + File.separator + "misc"
+            + File.separator + BINDING_ID;
+    private static final String DATABASE_FILES = ".json";
+
+    private final Logger logger = LoggerFactory.getLogger(MiIoDatabaseWatchService.class);
+    private Map<String, URL> databaseList = new HashMap<>();
+
+    @Activate
+    public MiIoDatabaseWatchService() {
+        super(LOCAL_DATABASE_PATH);
+        processWatchEvent(null, null, Paths.get(LOCAL_DATABASE_PATH));
+        populateDatabase();
+        if (logger.isTraceEnabled()) {
+            for (String device : databaseList.keySet()) {
+                logger.trace("Device: {} using URL: {}", device, databaseList.get(device));
+            }
+        }
+    }
+
+    @Override
+    protected boolean watchSubDirectories() {
+        return true;
+    }
+
+    @Override
+    protected WatchEvent.Kind<?>[] getWatchEventKinds(@Nullable Path directory) {
+        return new WatchEvent.Kind<?>[] { ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY };
+    }
+
+    @Override
+    protected void processWatchEvent(@Nullable WatchEvent<?> event, WatchEvent.@Nullable Kind<?> kind,
+            @Nullable Path path) {
+        if (path != null && path.getFileName().toString().endsWith(DATABASE_FILES)) {
+            logger.debug("Local Databases file {} changed. Refreshing device database.", path.getFileName());
+            populateDatabase();
+        }
+    }
+
+    /**
+     * Return the database file URL for a given modelId
+     *
+     * @param modelId the model
+     * @return URL with the definition for the model
+     */
+    public @Nullable URL getDatabaseUrl(String modelId) {
+        if (databaseList.containsKey(modelId)) {
```

Comment:
`get` returns a null-value if the key is not present, so you can omit the check and directly return the result from `get`. I wonder if it would not be better to parse the files directly here and return the desired information instead of the URL (i.e. move the parsing part out of the thing handler). This way the JSON parsing would only be done once.

Reply:
I can do the parsing here, but the main reason to store only the URL results of the parsing is that I expect the majority of users don't need the full files parsed. dunno what happens once we have a large amount of db files (there are over 500 devices nowadays in the Xiaomi ecosystem) . Seems like a waste of memory to store it fully in memory

Reply:
indeed, I removed now the if and directly return the result of the `get`